scripts/install_avalanchego_release.sh

```diff
@@ -11,41 +11,119 @@ source "$SUBNET_EVM_PATH"/scripts/versions.sh
-mkdir -p $AVALANCHEGO_BUILD_PATH
+BUILD_DIR=${AVALANCHEGO_BUILD_PATH}-${AVALANCHEGO_VERSION}
 
-if [[ ! -f ${AVAGO_DOWNLOAD_PATH} ]]; then
-  echo "downloading avalanchego ${VERSION} at ${AVAGO_DOWNLOAD_URL} to ${AVAGO_DOWNLOAD_PATH}"
-  curl -L ${AVAGO_DOWNLOAD_URL} -o ${AVAGO_DOWNLOAD_PATH}
-fi
-echo "extracting downloaded avalanchego to ${AVALANCHEGO_BUILD_PATH}"
-if [[ ${GOOS} == "linux" ]]; then
-  mkdir -p ${AVALANCHEGO_BUILD_PATH} && tar xzvf ${AVAGO_DOWNLOAD_PATH} --directory ${AVALANCHEGO_BUILD_PATH} --strip-components 1
-elif [[ ${GOOS} == "darwin" ]]; then
-  unzip ${AVAGO_DOWNLOAD_PATH} -d ${AVALANCHEGO_BUILD_PATH}
-  mv ${AVALANCHEGO_BUILD_PATH}/build/* ${AVALANCHEGO_BUILD_PATH}
-  rm -rf ${AVALANCHEGO_BUILD_PATH}/build/
+extract_archive() {
+  mkdir -p ${BUILD_DIR}
+
+  if [[ ${AVAGO_DOWNLOAD_PATH} == *.tar.gz ]]; then
+    tar xzvf ${AVAGO_DOWNLOAD_PATH} --directory ${BUILD_DIR} --strip-components 1
+  elif [[ ${AVAGO_DOWNLOAD_PATH} == *.zip ]]; then
+    unzip ${AVAGO_DOWNLOAD_PATH} -d ${BUILD_DIR}
+    mv ${BUILD_DIR}/build/* ${BUILD_DIR}
+    rm -rf ${BUILD_DIR}/build/
+  fi
+}
+
+# first check if we already have the archive
+if [[ -f ${AVAGO_DOWNLOAD_PATH} ]]; then
+  # if the download path already exists, extract and exit
+  echo "found avalanchego ${VERSION} at ${AVAGO_DOWNLOAD_PATH}"
+
+  extract_archive
+else
+  # try to download the archive if it exists
+  if curl -s --head --request GET ${AVAGO_DOWNLOAD_URL} | grep "302" > /dev/null; then
+    echo "${AVAGO_DOWNLOAD_URL} found"
+    echo "downloading to ${AVAGO_DOWNLOAD_PATH}"
+    curl -L ${AVAGO_DOWNLOAD_URL} -o ${AVAGO_DOWNLOAD_PATH}
+
+    extract_archive
+  else
+    # else the version is a git commitish (or it's invvalid)
+    GIT_CLONE_URL=https://github.com/ava-labs/avalanchego.git
+    GIT_CLONE_PATH=${BASEDIR}/avalanchego-repo/
+    
+    # check to see if the repo already exists, if not clone it 
+    if [[ ! -d ${GIT_CLONE_PATH} ]]; then
+      echo "cloning ${GIT_CLONE_URL} to ${GIT_CLONE_PATH}"
+      git clone --no-checkout ${GIT_CLONE_URL} ${GIT_CLONE_PATH}
+    fi
+
+    # check to see if the commitish exists in the repo
+    WORKDIR=$(pwd)
+
+    cd ${GIT_CLONE_PATH}
+
+    git fetch
+
+    echo "checking out ${AVALANCHE_GO_VERSION}"
+
+    set +e
+    # try to checkout the branch
+    git checkout origin/${AVALANCHE_GO_VERSION} > /dev/null 2>&1
+    CHECKOUT_STATUS=$?
+    set -e
+
+    # if it's not a branch, try to checkout the commit 
+    if [[ $CHECKOUT_STATUS -ne 0 ]]; then
+      set +e
+      git checkout ${AVALANCHE_GO_VERSION} > /dev/null 2>&1
+      CHECKOUT_STATUS=$?
+      set -e
+
+      if [[ $CHECKOUT_STATUS -ne 0 ]]; then
+        echo
+        echo "'${VERSION}' is not a valid release tag, commit hash, or branch name"
+        exit 1
+      fi
+    fi
+
+    COMMIT=$(git rev-parse HEAD)
+
+    # use the commit hash instead of the branch name or tag
+    BUILD_DIR=${AVALANCHEGO_BUILD_PATH}-${COMMIT}
```

Comment:
Hmm at first I thought that the below code where it builds iff the build dir does not exist may run into issues for branches if there were newly pushed changes, but I see that this ensures that the build dir is based off of the commit, so that's not an issue.

Reply:
I'm brilliant, what can I say, 😉